downloader/src/main/kotlin/Downloader.kt

```diff
@@ -131,7 +131,24 @@ class Downloader {
 
         try {
             if (target.vcsProcessed.url.isBlank()) {
-                throw DownloadException("No VCS URL provided for '${target.id}'.")
+                val details = " " + when (target.id.provider) {
+                    "Bundler" -> "Please define the \"source_code_uri\" in the \"metadata\" of the Gemspec, see: " +
+                            "https://guides.rubygems.org/specification-reference/#metadata"
+                    "Gradle" -> "Please make sure the release POM file includes the SCM connection, see: " +
+                            "https://docs.gradle.org/current/userguide/publishing_maven.html#" +
+                            "example_customizing_the_pom_file"
+                    "Maven" -> "Please define the <scm><connection> in the POM file, see: " +
+                            "http://maven.apache.org/pom.html#SCM"
+                    "NPM" -> "Please define the \"repository\" in the package.json file, see: " +
+                            "https://docs.npmjs.com/files/package.json#repository"
+                    "PIP", "PyPI" -> "Please make sure the setup.py defines the 'Source' attribute in " +
+                            "'project_urls', see: " +
+                            "https://packaging.python.org/guides/distributing-packages-using-setuptools/#project-urls"
+                    "SBT" -> "Please make sure the released POM file includes the SCM connection, see: " +
+                            "http://maven.apache.org/pom.html#SCM"
+                    else -> ""
```

Comment:
So in this case still " " gets prepended. Wouldn't it be nicer if `details` was really empty in this case? I.e., I'd suggest to inline the " " before all details strings, or do something like

    val detailsMessage = if (details.isNotEmpty()) " $details" else details
    throw DownloadException("No VCS URL provided for '${target.id}'.$detailsMessage")

Reply:
I have inlined the space. I also would have liked to use the package managers here, but the `downloader` is a dependency of the `analyzer`, so we do not know about package managers here. We would have to move the package managers to a separate module to make this possible, and this seems overhead for this change.

Reply:
Which reminds me: The only thing from the `downloader` module the `analyzer` depends on is the VCS classes. As these are growing more complex, I'm thinking about moving these to their own module, or at least utility package. But that's a different topics.